Fix #6383: Added regex to validate return_url in signup page

NishealJ opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Fixes #6383 : To patch the open url redirection in signup page, i've added regex to validate return_url.
Now, https://www.oppia.org/signup?return_url=https://google.com  will redirect to '/'
and https://www.oppia.org/signup?return_url=/hello will redirect = '/hello'
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

NishealJ committed:
Added Regex to validate return_url

NishealJ committed:
Added Regex to validate return_url

NishealJ commented:
Hi,
@seanlip @nithusha21  @1995YogeshSharma  PTAL. Thanks !

abhaygupta97 committed:
Fix #6370: Fix Title and Goal label alignment (#6372)

seanlip commented on core/controllers/profile.py:
```diff
@@ -251,7 +251,8 @@ class SignupPage(base.BaseHandler):
         """Handles GET requests."""
         return_url = str(self.request.get('return_url', self.request.uri))
-
+        if re.match('^/[^//]', return_url) is None:
```
I don't think I understand this pattern. Could you explain it?

Also, missing backend tests.

NishealJ replied:
Hi,

`^/[^//]` This pattern will only accept those URLs which starts from '/' and doesn't starts with  '//'
Example : '/hello' will be accepted and '//hello' will be rejected.

I'm rejecting the URLs starting with '//' because they are the bypass for URL Redirection 
`oppia.org/signup?return_url=//hello.com` will redirect to `hello.com` &
`oppia.org/signup?return_url=/hello.com` will redirect to `oppia.org/hello.com`

Also , all the URLs starting with https or http will also be rejected and if no match is found with the above pattern the site will be redirectd to '/' which is the learnerdashboard.

I feel its a good practice to redirect to internal directories '/something' than redirecting to https | http URLs need you views on the same :)

I'll run the backend tests.

Thanks

seanlip replied:
> I'll run the backend tests.

Just to be clear, I meant **writing** backend tests to ensure that your fix achieves the behaviour you mention.

Also, redirecting without https/http SGTM. Thanks!

NishealJ replied:
I'll write the backend tests and update you on this with the converge reports, hope to meet the desired results Thanks !

NishealJ replied:
Hi @seanlip ,

I've a doubt in writing tests for this, i'm matching the current host address with the redirected host address to find that the redirection is in same host address.or not.
will it be a good approach to write a testcase for the above problem?

seanlip replied:
I think you can just check that different return URLs that exemplify the different cases go to the right places.

NishealJ replied:
Yes, This can also be done.

seanlip commented on core/controllers/profile.py:
```diff
@@ -13,7 +13,7 @@
 
 """Controllers for the profile page."""
-
+import re
```
Add a newline above this one, and add a newline below this one. When coding make sure to follow existing styles (e.g. in other files).

NishealJ replied:
Definitely, i'll update this in next commit.

NishealJ replied:
Done Thanks !

DubeySandeep committed:
Added sean to manifest.json (#6421)

seanlip committed:
Make new model IDs only contain alphanumeric characters. (#6408)

* Make new model IDs only contain alphanumeric characters.

* Add test for alphanumeric chars.

jacoblipech committed:
Fix part of #5134: Make test coverage of core.controllers.collection_view 100% (#6427)

Fix part of #5134: Make test coverage of core.controllers.collection_viewer 100%

ankita240796 committed:
Fix part of #6343: Adds custom type definitions (#6430)

* Update typings

* Remove unused vars, make naming consistent

vinitamurthi committed:
Fix #6377: Create Practice Sessions Page Skeleton (#6381)

* Create Practice Sessions Page Skeleton

* Deindent brace by 4 spaces

NishealJ committed:
Fix Mistakes

NishealJ committed:
Fix Mistakes

abhaygupta97 committed:
Fixes part of #5002: Remove logoutUrl from HTML (#6401)

* Fixes part of #5002: Remove logoutUrl from HTML

* Adding urls to profile.py

* Fixing mistakes

* Making suggested changes

* Remove logout url

* Making changes

* Making requested changes

* Removing str from Logout page

anubhavsinha98 committed:
Fix #6100 : Added verbose and non verbose mode in the pre submit check (#6146)

* First

* First

* Second

* Third

* Fourth

* Added Mode

* Added verbose and non verbose mode

* Default non verbose mode

* Changed the arg mode to flag

* Changed the arg mode to flag

* Added -v flag

* removed print statement

* Update pre_commit_linter.py

* Update pre_commit_linter.py

* Changes done-global added

* Fixed changes

* Changes done

Added space after #

* Removed spaces

* Added class

* Added the LintsCheckManager

* Added functions as methods

* Moved to class

* Completed and Fixed Linting issues

* Added comments

* Changes Done

* Bool var name changed and few changes done

* Changes Done

* Changes Done

* changes done

* Made the changes

* Changes Done

* Changes Done

* Done the changes

* Removed

* Added verbose flag to make Travis build unstale

* Added 4 spaces

* Added filenames

* Added filepath

* added filepath

mzaman07 committed:
Fix #6248 : Prevent guest and unauthorized users from drag-and-drop audio in translation tab. (#6304)

amey-kudari committed:
Fix #6325 : modify style of active and inactive submenus on learner dashboard (#6358)

* added background and a subtle margin to selected li

* added background and a subtle margin to selected li

* modify color to hexa value, select color: blue

anubhavsinha98 committed:
Fixes #6205 : Disabled the reject button, enabled only when some review is entered. (#6384)

* First

* First

* Second

* Third

* Fourth

* Bug fixed

* Tooltip

* Added the message

import-keshav committed:
Fixes  #6316: Refactor base.html (#6399)

* remove topics,skill and story creation service from base.html

* remove ConstructTranslationIdsService from base.html

* remove StateRulesStatsService

seanlip committed:
Reorder CODEOWNERS to ensure that the most important rules are at the end. (#6438)

DubeySandeep committed:
Added filter filer in exploration_player. (#6437)

brianrodri committed:
Fix #5802, #5590: Introduce Playthrough Improvements Cards (#6400)

* initial commit

* introduce ImprovementCardService

* add initial (failing) tests

* add minimal code to get ICS to pass

* improve docstring

* use done to test async properly

* force fetchCards to return empty array

* better name

* move helper var into beforeEach

* comment that ordering isn't important

* add . prefix

* better describe string

* more context in describe string

* consolidate expected factories

* widen contract for improvement cards

* use resolutionActions

* test resolution actions

* simplify order check

* s/resolve/perform

* better it description

* add TODOs to return values

* s/"/'

* factor out actions into a domain object (with tests)

* fix indent errors

* fix indents

* better naming

* break list into lines

* better names

* fix fileoverviews

* add check to see registry is frozen

* use ImprovementCardService to render the cards in the improvements tab

* remove console message

* place buttons in div

* s/ImprovementResolutionAction/ImprovementAction

* compact lines

* compact lines

* add todo

* add todo

* add clause to translate action names

* use 'Archive' as button text

* s/Archive Card/Archive

* improved JSDoc

* build data from playthrough issues

* use variables to document args

* return promise from resolve

* return value of action

* test http callout

* better naming

* always return a promise

* always return promises

* cleaner checks

* better declaration organization

* fix post expectations

* simpler organization

* better binding to improvements tab

* cache fetched cards

* fix usage of cache

* fix indents

* fix comma-spacing

* move openPlaythroughModal to PlaythroughIssuesService

* dedupe modal logic with PIS.showPlaythroughModal

* s/show/open

* simplify condition

* use directives to render card data

* fix errors for rendering as directive

* add missing closing tag

* cleaner copy of base directive

* dedupe class attribute

* fix lint errors

* add missing semi-colon

* add copyright

* s/getActions/getActionButtons

* address comments

* s/performAction/execute

* remove misleading comment

* remove already tested function

* changes in async checks

* fix async tests

* better test organization

* fix indent

* execute done separately

* nit formatting change

* inline function

* s/name/text for action buttons

* address comments

* s/getName/getText

* use bidirectional binding

* capitalize JSDoc descriptions

* improve card formatting

* fix hexagon button

* only show open cards in summary

* UX improvements

* add generalized "confirmation modal"

* fix async tests

* better test coverage

* better test structure

* reorder tests

* remove comma

* use one-way reading

* add coverage for createNew

* test only title for identity

* match css to mock

* fix padding of containing div

* fix css

* remove unittest promise

* add expectation

* bigger font-size

* revert font-size change

* change font-size in HTML

* cache backend calls to playthrough issues

* bring back hexagon text class

* improve JSDoc

* use correct optional JSDoc notation

* remove .trustAsHtml

* use <[notation]>

* better description about getDirectiveData

* add type annotation

* capitalize description

* do not cache results in ICS

* remove cache usages

* more accurrate comment

* remove unused import

* add more unit test coverage for functions

* add docstring for param

* add docstring for param

* fix param name

* remove unused function

* swap implementation

* remove unused scope variables

* add brace scope

* improved var name

* better var naming

* s/resolved/open

* better punctuation

* invalidate cache after resolution

* remove i18n translations for outdated text

* remove i18n of improvements tab - we don't i18n the creator view

* perform cleaner cache refreshing after resolving issues

* remove redundant return

* better punctuation

* remove redundant .then

* use angular.equals

* better param name

* simpler param

seanlip committed:
Change the name of the schema_version field in story models. (#6420)

anubhavsinha98 committed:
Moving lint tests to CircleCI from Travis (#6453)

* First

* First

* Second

* Third

* Fourth

* Reset

* lint tests on circleci

NishealJ committed:
Fixes #6414: REQUIRE_PAYLOAD_CSRF_CHECK Flag Flipped in Error404Handler class. (#6435)

* REQUIRE_PAYLOAD_CSRF_CHECK Flag Flipped in Error404Handler class for Fix#6414

* Fix Backend Test v1.2

* Removing redundancy Adding Pass statement

BenHenning committed:
Update pip and Skulpt installations for newer versions of Ubuntu (#6464)

* Update pip and Skulpt installation to work properly on newer versions of
Ubuntu.

* Address reviewer comment: try both pip installs.

* Clarify the new skulpt sed commands.

* Rearrange comments and clarify what the new disable lines are doing per
code review comments.

jameesjohn committed:
Fixed part of #5002:  Removed iframed from GLOBALS (#6410)

* removed iframed from GLOBALS

* removed iframed from base.html in tests directory

Commit:
Adding backend tests

Commit:
Adding backend tests

anubhavsinha98 committed:
Fixes part of #5134 : Increasing the test coverage of subtopic_page_services.py from 64% to 100%. (#6360)

* First

* First

* Second

* Third

* Fourth

* 96% Covered

* 96%

* Fixed Linting Issues

* Fixed Linting Issues

* changed

* 98% and made more robust

* Done

* Added

* Added one condition

* Added append Noen

* Added more tests

* Added more tests

Commit:
backend test v1.1

Commit:
backend test v1.1

NishealJ commented:
Hi @seanlip , PTAL Thanks !

anubhavsinha98 committed:
Added webhook links for Travis, CircleCI & Codecov (#6398)

* First

* First

* Second

* Third

* Fourth

* Added Travis links

* webhook for circle ci

* Done

* Reset

* Added codecov yml file

* Done

* Deleted codecov.yml

Deleted codecov yml file

import-keshav committed:
Fix #6294: Upgrade BeautifulSoup version from 4.6.0 to 4.7.1 (#6355)

* upgrade beautiful soup libraries

* add soupsieve package

* leave it

* add backports.functools_lru_cache package

* fix some changes

seanlip commented on core/controllers/profile.py:
```diff
@@ -251,7 +253,8 @@ class SignupPage(base.BaseHandler):
         """Handles GET requests."""
         return_url = str(self.request.get('return_url', self.request.uri))
-
+        if re.match('^/[^//]', return_url) is None:
```
Add a line comment above this line that explains what it's checking for, for the benefit of devs reading this code (to give them context on what this check is looking for and what its aim is). So they don't remove it in the future.

NishealJ replied:
done

seanlip commented on core/controllers/profile_test.py:
```diff
@@ -45,6 +45,31 @@ def test_going_somewhere_else_while_signing_in_logs_user_out(self):
+        response = self.get_html_response(
+            '/signup?return_url=/page', expected_status_int=302)
+
+        self.assertIn('/', response.headers['location'])
```
I believe this is only checking it for the most recent response. You need to do the assertion immediately after you get the response in each case.

Also, check equality for the whole URL, don't just check containment. That would probably have caught the issue.

NishealJ replied:
Sorry Missed it !

Also, you mean to use assertEqual(‘https://localhost:8081’,response.headers[‘location’]) ?

I dint check the equality of whole url thinking in mind that other devs may use other hosts or ports which can cause issue.

seanlip replied:
You can strip the host/port part out, and check equality of the remaining string.

NishealJ replied:
Adding Third party libraries like urlparse for the same or regex only?
parsing such with urlparse will be very smooth but with urlparse we may find difficulties while migrating to python3.

seanlip replied:
No need to import third-party libraries.

NishealJ replied:
sure

abhaygupta97 committed:
Fixes part of #6387: Remove globals from library.html and preferences.html (#6415)

* Remove global SEARCH_DROPDOWN_CATEGORIES

* Remove global PAGE_MODE

* Remove global LANGUAGE_CODES_AND_NAMES

* Making suggested changes

* Moving variable to constants

* Adding requested changes

* Fixing incorrect code

* Fixing line alignment

* Making requested changes

* Adding review changes

anubhavsinha98 committed:
Fixes #6426: Added regex check for exploration id & nodes now can be saved with null exploration ids. (#6470)

* First

* First

* Second

* Third

* Fourth

* Reset

* Fixed bug: Added regex check and multiple chapters can have empty exp id

* Added exception

* Added comment

sandeep-patel11 committed:
Fixes part of #5898: Added Google Analytics for translation tab (#6166)

NishealJ committed:
Fix #6461: Changing error status for missing csrf_token from 500 to 401. (#6463)

* changing error status of missing csrf_token

* Changing error message

* Adding backend test for changed error status

* fix mistakes in error message

ankita240796 committed:
Fix server errors (#6454)

nithusha21 committed:
Automate changelog generation in the release_info script (#6395)

* add script to categorize PRs in the changelog based on the labels

* remove stray changes from debugging

* make review changes

* fix wording

* fix wording in print

Commit:
Fix #5426: Lint check to ensure that each Angular file contains exactly one service/directive/etc. inside it. (#6202)

* Check js file has only one property

* Check js file has only one property

* Check js file has only one property

* Refactor extensions/

* Resolved conflicts

* Fix errors

* Fix errors

* Fix errors

* Lint errors

* Lint errors

* Fix errors

* Fix errors

* Fix errors

* Fix errors

* Fix errors

* Fix base test

* admin-roles-tab

* Removed unneccessary scripts

* Change string patterns

* Fix lint errors

* Fix e2e tests

* Address review comments

* Addressed comments

* Addressed comments

* Final reviews

* Improve docstring

* Improved docstring

* addressed review comments

* lint checks

* Nit

* Resolve conflicts

* Minor nit

* Resolve errors

* Fix spaces

* Fix spaces

* Fix spaces

* Fix spaces

* Fix spaces

kevinlee12 committed:
Temporary code ownership update (#6476)

* Update codeownership

* Specify which dirs need to be reverted in CODEOWNERS

vojtechjelinek committed:
Add option not to start browser window to start.sh (#6363)

* add option not to start browser window

* remove mistake webpack line

* remove webpack

* add gulp back

bansalnitish committed:
Fixes #4632: Fixes learn again-continue button (#6448)

* Fix learn again button

* moved to previous line

NishealJ committed:
Adding comments in profile.py

NishealJ committed:
Adding comments in profile.py

NishealJ committed:
fixed mistakes in profile_test.py

NishealJ committed:
fixed mistakes in profile_test.py

NishealJ commented:
Hi @seanlip , PTAL Thanks !

seanlip commented on core/controllers/profile.py:
```diff
@@ -251,7 +253,9 @@ class SignupPage(base.BaseHandler):
         """Handles GET requests."""
         return_url = str(self.request.get('return_url', self.request.uri))
-
+        #validating return_url for no external redirection.
```
Follow existing codebase standards: start with a capital letter, add space after '#'.

```
# Validate return_url to prevent redirection to external websites.
```

/cc @oppia/dev-workflow-team Can any of these be turned into automatic checks?

lilithxxx replied:
There is a function for checking comments [here](https://github.com/oppia/oppia/blob/develop/scripts/pre_commit_linter.py#L1729). @nishealJ please update your branch with oppia/develop to run the lint tests on circle CI. Thanks

seanlip commented on core/controllers/profile_test.py:
```diff
@@ -45,6 +46,40 @@ def test_going_somewhere_else_while_signing_in_logs_user_out(self):
+        self.assertEqual('', strip_domain_from_location_header(
+            response.headers['location']))
+
+        response = self.get_html_response(
```
Good tests. Add a few more cases -- multiple parts in the URL e.g. /page/abc, query params, hash params, etc. to make sure those all get redirected successfully.

bansalnitish committed:
Fixes #4931: Fixes frozen editor tour (#6450)

* Fix frozen editor tour

* Fixed the fragile behavviour

* Changed to saveButtonTutorialElement

NishealJ committed:
Merge branch 'fix-6383' of https://github.com/nishealJ/oppia into fix-6383

NishealJ committed:
Mistakes fixed & Added more testcases

NishealJ committed:
Mistakes fixed & Added more testcases

vojtechjelinek commented:
@nishealJ You probably made some mistake when merging.

NishealJ commented:
Hi @vojtechjelinek,  yes a wrong merge I'll just revert and clean this PR.

NishealJ committed:
Revert "Merge branch 'fix-6383' of https://github.com/nishealJ/oppia into fix-6383"

This reverts commit 3c526fba4bd80e0f0eff8c65583aacb121677fe8, reversing
changes made to 94af7993dfb6ec02ab3a47e52291f0a036423de8.

NishealJ commented:
Hi @seanlip, 
I made some mistake while merging,  can I raise a new PR for the same?
Thanks

vojtechjelinek commented:
Yes, you can!

NishealJ commented:
Hi,
New PR raised for the same #6498.
Thanks
